Add NetWave integration

TheLogicMaster opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
```yaml
# Example configuration.yaml
camera:
  - platform: netwave
    address: http://ADDRESS:PORT
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [Documentation Pull Request](https://github.com/home-assistant/home-assistant.io/pull/13575)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

TheLogicMaster committed:
Add NetWave integration

homeassistant commented:
Hi @TheLogicMaster, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#36063).

Once you do that we will be able to review and accept this pull request.

Thanks!

TheLogicMaster committed:
Fixed typo in send_parameter service

frenck commented on homeassistant/components/netwave/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+                        CONF_TIMEOUT,
+                        default=user_input.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+                    ): int,
+                    vol.Required(
```
These settings should not be part of the integration setup, but part of an options flow.
The configuration flow should only contain the bare necessity to set up an integration.

For more information about configuration options:

<https://developers.home-assistant.io/docs/config_entries_options_flow_handler/>

frenck commented on homeassistant/components/netwave/__init__.py:
```diff
@@ -0,0 +1,211 @@
+    hass.services.async_register(
+        DOMAIN, SERVICE_COMMAND, async_handle_command, schema=SERVICE_COMMAND_SCHEMA
+    )
+    hass.services.async_register(
```
These services can be greatly simplified when they are set up as entity services.

For more information see here: <https://developers.home-assistant.io/docs/dev_101_services#entity-services>

frenck commented on homeassistant/components/netwave/__init__.py:
```diff
@@ -0,0 +1,211 @@
+    """Unload camrta config entry."""
+    await hass.config_entries.async_forward_entry_unload(entry, "camera")
+
+    _LOGGER.error(hass.data[DOMAIN])
+    _LOGGER.error(entry.unique_id)
```
Stale debug strings?

frenck commented on homeassistant/components/netwave/__init__.py:
```diff
@@ -0,0 +1,211 @@
+        _LOGGER.error(camera.unique_id)
+        if camera.unique_id == entry.unique_id:
+            hass.data[DOMAIN].pop(camera.entity_id)
+            _LOGGER.error(camera.entity_id)
```
Stale debug log?

frenck commented:
Hi there @TheLogicMaster!

Added a couple of comments to get started.

Additionally, please note, it is required to have tests for the configuration flow with 100% coverage. Please add those as well.

TheLogicMaster committed:
NetWave unit tests, entity services, and config-flow options

TheLogicMaster committed:
Fixed NetWave test requirements

TheLogicMaster commented:
Hi, I moved the unnecessary config-flow options to a second step and added an options flow for those secondary options. All of the default values are auto-filled in that second step, but it could just be removed and delegated to the options flow if that suits the standard better.

TheLogicMaster committed:
Fixed NetWave lint issues

TheLogicMaster committed:
Fixed NetWave return-after-else lint issue

frenck commented:
Not stale

frenck commented on .coveragerc:
```diff
@@ -479,6 +479,7 @@ omit =
     homeassistant/components/netgear/device_tracker.py
     homeassistant/components/netgear_lte/*
     homeassistant/components/netio/switch.py
+    homeassistant/components/netwave/*
```
`config_flow.py` should not be excluded from the coverage.
